// arcjet/lib/src/store.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const readline = require("readline");
const js_sha3_1 = require("js-sha3");
const assert_1 = require("assert");
const utils_1 = require("./utils");
class Store {
    constructor() {
        this.positions = {};
        this.lengths = {};
        this.dblen = 0;
        this.shaLength = 64;
        this.open = (path = this.path) => __awaiter(this, void 0, void 0, function* () { return yield utils_1.open(path, 'a+'); });
        this.close = (fd) => __awaiter(this, void 0, void 0, function* () { return yield utils_1.close(fd); });
        this.init = (path) => new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            try {
                this.path = path;
                const instream = fs.createReadStream(path, 'utf8');
                const rl = readline.createInterface(instream);
                rl.on('line', line => {
                    const hash = line.substr(0, this.shaLength);
                    this.update(hash, line.length - 1 - this.shaLength);
                });
                rl.on('close', () => {
                    resolve();
                });
                rl.on('error', err => {
                    reject(err);
                });
            }
            catch (err) {
                console.error(err);
                reject(err);
            }
        }));
    }
    update(hash, length) {
        this.positions[hash] = this.dblen;
        this.lengths[hash] = length;
        this.dblen = this.dblen + this.shaLength + 1 + length + 1;
    }
    set(data) {
        return __awaiter(this, void 0, void 0, function* () {
            assert_1.strict.ok(typeof data === 'string');
            assert_1.strict.ok(data.length > 0);
            const hash = js_sha3_1.sha3_256(data);
            if (this.lengths[hash] && this.lengths[hash] > 0) {
                return hash;
            }
            const fd = yield this.open();
            yield utils_1.appendFile(fd, `${hash} ${data}\n`, 'utf8');
            this.update(hash, data.length);
            yield this.close(fd);
            return hash;
        });
    }
    get(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            assert_1.strict.ok(typeof hash === 'string');
            assert_1.strict.ok(hash.length === 64);
            const position = this.positions[hash];
            const length = this.lengths[hash];
            const readLength = this.shaLength + 1 + length;
            if (!position && !length) {
                return {
                    hash,
                    data: undefined,
                    error: 'NOTFOUND',
                };
            }
            const fd = yield this.open();
            const { buffer } = yield utils_1.read(fd, Buffer.alloc(readLength, 0, 'utf8'), 0, readLength, position);
            yield this.close(fd);
            const result = buffer.toString('utf8');
            const [resultHash, rawData] = result.split(' ');
            return {
                hash: resultHash,
                data: rawData,
            };
        });
    }
}
exports.default = Store;
