// arcjet/lib/client.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const nacl = require("tweetnacl");
const QRCode = require("qrcode");
const parser_1 = require("./parser");
const client_utils_1 = require("./client_utils");
const types_1 = require("./types");
const hashAsByteArray = (data) => nacl.hash(client_utils_1.hexToBytes(data));
const hashAsString = (data) => client_utils_1.bytesToHex(hashAsByteArray(data));
class Arcjet {
    constructor(host = 'http://127.0.0.1:3000') {
        this.owners = {};
        this.shaLength = 128;
        this.validate = (record) => __awaiter(this, void 0, void 0, function* () {
            const { recordHash, ownerHash, dataHash, signature, data } = parser_1.parseRecord(record);
            let ownerPublicKey = this.owners[ownerHash];
            if (!ownerPublicKey) {
                const req = yield fetch(`${this.host}/store/${ownerHash}`);
                const data = yield req.text();
                ownerPublicKey = parser_1.parseRecord(data).data; // TODO validate this record too
            }
            const recDataHash = hashAsString(data);
            const recRecordHash = hashAsString(record.substr(this.shaLength + 1));
            const verified = nacl.sign.detached.verify(client_utils_1.hexToBytes(recDataHash), client_utils_1.hexToBytes(signature), client_utils_1.hexToBytes(ownerPublicKey));
            console.log(verified, recDataHash === dataHash, recRecordHash === recordHash, recRecordHash, recordHash);
            if (verified && recDataHash === dataHash && recRecordHash === recordHash) {
                return data;
            }
            else {
                return '400';
            }
        });
        this.defaultOwnerHash = '0'.repeat(this.shaLength);
        this.host = host;
    }
    getCurrentParentHash(owner) {
        return __awaiter(this, void 0, void 0, function* () {
            const req = yield fetch(`${this.host}/parent/${owner}`);
            const data = yield req.text();
            return data;
        });
    }
    download(data) {
        var link = document.createElement('a');
        link.download = 'filename.png';
        link.href = data;
        link.click();
    }
    save(keys) {
        return __awaiter(this, void 0, void 0, function* () {
            localStorage.setItem(types_1.ArcjetStorageKeys.ARCJET_PUBLIC_KEY, keys.publicKey);
            localStorage.setItem(types_1.ArcjetStorageKeys.ARCJET_PRIVATE_KEY, keys.privateKey);
            localStorage.setItem(types_1.ArcjetStorageKeys.ARCJET_OWNER_HASH, this.defaultOwnerHash);
            const ownerHash = yield this.set(keys.publicKey, 'owner');
            localStorage.setItem(types_1.ArcjetStorageKeys.ARCJET_OWNER_HASH, ownerHash);
        });
    }
    load() {
        const ARCJET_OWNER_HASH = localStorage.getItem(types_1.ArcjetStorageKeys.ARCJET_OWNER_HASH);
        const ARCJET_PUBLIC_KEY = localStorage.getItem(types_1.ArcjetStorageKeys.ARCJET_PUBLIC_KEY);
        const ARCJET_PRIVATE_KEY = localStorage.getItem(types_1.ArcjetStorageKeys.ARCJET_PRIVATE_KEY);
        if (ARCJET_OWNER_HASH && ARCJET_PUBLIC_KEY && ARCJET_PRIVATE_KEY) {
            return {
                ARCJET_OWNER_HASH,
                ARCJET_PUBLIC_KEY,
                ARCJET_PRIVATE_KEY,
            };
        }
        else {
            throw 'No Auth Data';
        }
    }
    owner() {
        return localStorage.getItem(types_1.ArcjetStorageKeys.ARCJET_OWNER_HASH);
    }
    generate() {
        return __awaiter(this, void 0, void 0, function* () {
            try {
                const keys = nacl.sign.keyPair();
                const privateKey = client_utils_1.bytesToHex(keys.secretKey);
                const publicKey = client_utils_1.bytesToHex(keys.publicKey);
                const qr = yield QRCode.toDataURL(privateKey);
                if (document)
                    this.download(qr);
                if (localStorage)
                    yield this.save({ privateKey, publicKey });
            }
            catch (err) {
                console.error(err);
            }
        });
    }
    get(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            const res = yield fetch(`${this.host}/store/${hash}`);
            if (res.status === 200) {
                const record = yield res.text();
                return yield this.validate(record);
            }
            return '404';
        });
    }
    set(data, tag, encoding = 'utf-8', type = 'text/plain') {
        return __awaiter(this, void 0, void 0, function* () {
            const { ARCJET_OWNER_HASH: ownerHash, ARCJET_PRIVATE_KEY: privateKey, } = this.load();
            const dataHashArray = hashAsByteArray(data);
            let parentHash;
            if (ownerHash === this.defaultOwnerHash) {
                parentHash = this.defaultOwnerHash;
            }
            else {
                parentHash = yield this.getCurrentParentHash(ownerHash);
            }
            const signature = nacl.sign.detached(dataHashArray, client_utils_1.hexToBytes(privateKey));
            const record = [
                ownerHash,
                parentHash,
                client_utils_1.bytesToHex(dataHashArray),
                encoding.padEnd(32, ' '),
                type.padEnd(32, ' '),
                tag.padEnd(32, ' '),
                client_utils_1.bytesToHex(signature),
                data,
            ].join('\t');
            const recordHash = hashAsString(record);
            const recordString = [recordHash, record].join('\t');
            const res = yield fetch(`${this.host}/store`, {
                method: 'POST',
                body: recordString,
            });
            const recRecordHash = yield res.text();
            client_utils_1.assert(recordHash === recRecordHash, 'Record hash from server must match computed record hash');
            return recordHash;
        });
    }
    findByTag(ownerHash, tag, limit, offset) {
        return __awaiter(this, void 0, void 0, function* () {
            const url = [this.host, 'find', ownerHash, tag];
            if (limit)
                url.push(limit.toString());
            if (limit && offset)
                url.push(offset.toString());
            const res = yield fetch(url.join('/'));
            if (res.status === 200) {
                const response = yield res.text();
                const records = response.split('\n');
                const results = yield Promise.all(records.map(this.validate));
                return results;
            }
            return ['404'];
        });
    }
}
exports.default = Arcjet;
