// arcjet/lib/store.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = require("fs");
const readline = require("readline");
const assert_1 = require("assert");
const nacl = require("tweetnacl");
const utils_1 = require("./utils");
const client_utils_1 = require("./client_utils");
const parser_1 = require("./parser");
const hashAsByteArray = (data) => nacl.hash(client_utils_1.hexToBytes(data));
const hashAsString = (data) => client_utils_1.bytesToHex(hashAsByteArray(data));
class Store {
    constructor() {
        this.positions = {};
        this.lengths = {};
        this.owners = {};
        this.dblen = 0;
        this.shaLength = 128;
        this.open = (path = this.path) => __awaiter(this, void 0, void 0, function* () { return yield utils_1.open(path, 'a+'); });
        this.close = (fd) => __awaiter(this, void 0, void 0, function* () { return yield utils_1.close(fd); });
        this.init = (path) => new Promise((resolve, reject) => __awaiter(this, void 0, void 0, function* () {
            this.emptyHash = '0'.repeat(this.shaLength);
            try {
                this.path = path;
                const instream = fs.createReadStream(path, 'utf8');
                const rl = readline.createInterface(instream);
                rl.on('line', line => {
                    const hash = line.substr(0, this.shaLength);
                    const ownerHash = line.substr(this.shaLength + 1, this.shaLength);
                    this.update(hash, line.length + 1, ownerHash);
                });
                rl.on('close', () => {
                    resolve();
                });
                rl.on('error', err => {
                    reject(err);
                });
            }
            catch (err) {
                console.error(err);
                reject(err);
            }
        }));
    }
    update(hash, length, ownerHash) {
        this.positions[hash] = this.dblen;
        this.lengths[hash] = length;
        this.owners[ownerHash === this.emptyHash ? hash : ownerHash] = hash;
        this.dblen = this.dblen + length;
        console.log('hash', hash, 'ownerHash', ownerHash, 'length', length);
    }
    set(record) {
        return __awaiter(this, void 0, void 0, function* () {
            const { recordHash, ownerHash, parentHash, dataHash, encoding, type, tag, signature, data, } = parser_1.parseRecord(record);
            assert_1.strict.ok(typeof data === 'string', 'Data must be of type string for now');
            assert_1.strict.ok(data.length > 0, 'Data must not be empty');
            assert_1.strict.ok(data.length <= 1000000000, 'Data must not be larger than 1GB');
            assert_1.strict.ok(data.includes('\t') === false, 'Data must encode all tabs');
            assert_1.strict.ok(data.includes('\n') === false, 'Data must encode all newlines');
            assert_1.strict.ok(ownerHash.length === this.shaLength, 'Supplied Owner Hash invalid');
            assert_1.strict.ok(encoding.length <= 32, 'Encoding length must be less than or equal to 32 characters');
            assert_1.strict.ok(type.length <= 32, 'Type length must be less than or equal to 32 characters');
            assert_1.strict.ok(tag.length <= 32, 'Tag length must be less than or equal to 32 characters');
            assert_1.strict.ok(tag.length > 0, 'Tag must be provided');
            assert_1.strict.ok(hashAsString(data) === dataHash, 'dataHash must be valid');
            assert_1.strict.ok(parentHash === this.getCurrentParent(ownerHash), 'parentHash CAS error');
            const verifiedRecord = [
                ownerHash,
                parentHash,
                dataHash,
                encoding.padEnd(32, ' '),
                type.padEnd(32, ' '),
                tag.padEnd(32, ' '),
                signature,
                data,
            ].join('\t');
            const verifiedRecordHash = hashAsString(verifiedRecord);
            if (this.lengths[verifiedRecordHash] &&
                this.lengths[verifiedRecordHash] > 0) {
                return verifiedRecordHash;
            }
            const recordString = [verifiedRecordHash, verifiedRecord].join('\t') + '\n';
            this.update(recordHash, recordString.length, ownerHash);
            const fd = yield this.open();
            yield utils_1.appendFile(fd, recordString, 'utf8');
            yield this.close(fd);
            return recordHash;
        });
    }
    get(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            assert_1.strict.ok(typeof hash === 'string', 'record hash must be a string');
            assert_1.strict.ok(hash.length === this.shaLength, 'record hash must be 128 characters in length');
            const position = this.positions[hash];
            const length = this.lengths[hash];
            if (!position && !length) {
                throw 'Record Not Found';
            }
            let recordBuffer = Buffer.alloc(length, 'utf8');
            const fd = yield this.open(this.path);
            const { buffer } = yield utils_1.read(fd, recordBuffer, 0, length, position);
            const recordString = buffer.toString('utf8');
            return recordString;
        });
    }
    getStream(hash) {
        return __awaiter(this, void 0, void 0, function* () {
            assert_1.strict.ok(typeof hash === 'string', 'record hash must be a string');
            assert_1.strict.ok(hash.length === this.shaLength, 'record hash must be 128 characters in length');
            const position = this.positions[hash];
            const length = this.lengths[hash];
            if (!position && !length) {
                throw 'Record Not Found';
            }
            const stream = fs.createReadStream(this.path, {
                encoding: 'utf8',
                start: position,
                end: position + length - 2,
            });
            return stream;
        });
    }
    findByTag(ownerHash, tag, limit = 0, offset = 0) {
        return __awaiter(this, void 0, void 0, function* () {
            assert_1.strict.ok(typeof ownerHash === 'string', 'owner hash must be a string');
            assert_1.strict.ok(ownerHash.length === this.shaLength, 'owner hash must be 128 characters in length');
            assert_1.strict.ok(typeof tag === 'string', 'tag must be a string');
            assert_1.strict.ok(tag.length <= 32, 'tag must be 32 characters in length or less');
            let hash = this.owners[ownerHash];
            let position = this.positions[hash];
            let length = this.lengths[hash] - 1;
            const fd = yield this.open(this.path);
            const results = [];
            let currentLimit = 0;
            let currentOffset = 0;
            let go = true;
            while (hash !== this.emptyHash && go) {
                let recordBuffer = Buffer.alloc(length, 'utf8');
                const { buffer } = yield utils_1.read(fd, recordBuffer, 0, length, position);
                const recordString = buffer.toString('utf8');
                const record = parser_1.parseRecord(recordString);
                hash = record.parentHash;
                position = this.positions[hash];
                length = this.lengths[hash] - 1;
                if (tag === record.tag) {
                    currentOffset++;
                    if (offset === 0 || currentOffset > offset) {
                        if (limit === 0 || currentLimit <= limit) {
                            results.push(recordString);
                            currentLimit++;
                        }
                    }
                }
                if (limit !== 0 && currentLimit >= limit) {
                    go = false;
                }
            }
            yield this.close(fd);
            return results.join('\n');
        });
    }
    getCurrentParent(ownerHash) {
        return this.owners[ownerHash] || this.emptyHash;
    }
}
exports.default = Store;
